#15 Announced Transfer and Aggregate Transactions using Catbuffer

fullcircle23 opened the pull request:
Java SDK now can announce TransferTransaction and AggregateTransaction using Catbuffer instead of FlatBuffer in Cow network. However, there is an issue with the size whereby it's 1 byte less than when using FlatBuffer but this does not seem to affect Announcing, Confirmation or Getting the Completed Transactions from the network via REST API. 
Closes #15
Next steps (to open new issue):
1. Find root cause of the difference in 1 byte.
In parallel:
2a) Apply the FlatBuffer to Catbuffer conversion patterns to other Transaction types.
2b) Upgrade to Catbuffer for Dragon network.

rg911 commented on src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java:
```diff
@@ -557,7 +561,15 @@ void aggregateSecretProofTransaction() throws ExecutionException, InterruptedExc
+        Observable<Transaction> observable = listener.confirmed(address);
+        observable = observable.take(1);
+        Future<Transaction> future = observable.toFuture();
+        while (!future.isDone()) {
```
Make sure to have timeout setting for the integration test. To get the confirmed message can take a while. 

Also need to subscribe to the `status` channel for any errors, incase the tx doesn't pass through. 

e.g. `@Test(timeout=90000)`

fullcircle23 replied:
Noted.

fullcircle23 replied:
Added Timeout when getting Future result, TimeoutException and a variable to easily modify timeout value, to resolve this issue.

rg911 commented on src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java:
```diff
@@ -557,7 +561,15 @@ void aggregateSecretProofTransaction() throws ExecutionException, InterruptedExc
+        observable = observable.take(1);
+        Future<Transaction> future = observable.toFuture();
+        while (!future.isDone()) {
+            System.out.println("not done, wait 1 second...");
```
again, don't think we need println

fullcircle23 replied:
yes. will remove it

fullcircle23 replied:
Done

rg911 commented on src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java:
```diff
@@ -557,7 +561,15 @@ void aggregateSecretProofTransaction() throws ExecutionException, InterruptedExc
+        Future<Transaction> future = observable.toFuture();
+        while (!future.isDone()) {
+            System.out.println("not done, wait 1 second...");
+            Thread.sleep(1000);
```
probably don't need it at all?

fullcircle23 replied:
for debugging only. i'll be removing them.

fullcircle23 replied:
Done

rg911 commented on src/main/java/io/nem/catapult/builders/KeyDto.java:
```diff
@@ -21,13 +21,15 @@
 /** Key. */
 public final class KeyDto {
     /** Key. */
+    public final static int SIZE = 32;
```
not sure we should change the generated catbuffer code or not??

fullcircle23 replied:
Yes. We can make improvements that to generated catbuffer code which can be feasibly ported back to the Catbuffer Java code generator

rg911 commented on src/main/java/io/nem/catapult/builders/SignatureDto.java:
```diff
@@ -48,13 +50,22 @@ public SignatureDto(final ByteBuffer signature) {
         }
     }
 
+    /**
```
same above

rg911 commented on src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java:
```diff
@@ -80,10 +85,89 @@ void serialization() {
+
+        assertEquals(txModel.getRecipient().plain(), txBuilder.getRecipient().asString());
+
+        // assert
```
remove the commented code?

fullcircle23 replied:
yes. can be removed

fullcircle23 replied:
Done

rg911 commented on src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java:
```diff
@@ -80,10 +85,89 @@ void serialization() {
+        System.out.println("\nTransactionBuilder asString:");
+        System.out.println(txBuilder.asString());
+
+        System.out.println("\nTransactionBuilder each field:");
```
remove the debug println?

fullcircle23 replied:
will be keeping some of the debug println until fully implementation of catbuffer

fullcircle23 commented on src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java:
```diff
@@ -63,44 +66,45 @@ void setup() throws ExecutionException, InterruptedException, IOException {
         TransactionAnnounceResponse transactionAnnounceResponse = transactionHttp.announce(signedTransaction).toFuture().get();
         assertEquals("packet 9 was pushed to the network via /transaction", transactionAnnounceResponse.getMessage());
 
+        System.out.println("signedTransaction hash: " + signedTransaction.getHash());
```
I'll remove the println. It's there for debugging only.

fullcircle23 replied:
Added Timeout when getting Future result, TimeoutException and a variable to easily modify timeout value, to resolve this issue.

fullcircle23 replied:
Done

fullcircle23 commented:
Fixed the difference in 1 byte which was due to missing message type prefix in message payload. Added Timeout when waiting for transaction complete validation result in E2E tests. Cleanup debug lines.
